ZenodoExportPlugin.php

```diff
@@ -134,56 +159,107 @@ public function getSettingsFormClassName(): string
+            return [['plugins.importexport.zenodo.api.error.noDoi']];
         }
 
-        if (!$mintDoi && !$object->getCurrentPublication()->getDoi()) {
-            return [['plugins.importexport.zenodo.register.error.noDoi']];
+        $zenodoApiUrl = ($this->isTestMode($context) ? self::ZENODO_API_URL_DEV : self::ZENODO_API_URL);
+        $recordsApiUrl = $zenodoApiUrl . self::ZENODO_API_OPERATION;
+
+        $isUpdate = false;
+        $isPublished = false;
+        if ($existingZenodoRecId = $object->getData($this->getIdSettingName())) {
+            $isUpdate = true;
+            $isPublished = $this->isRecordPublished($existingZenodoRecId, $recordsApiUrl);
         }
 
-        $headers = [
-            'Accept' => 'application/json',
-            'Content-Type' => 'application/json',
-            'Authorization' => 'Bearer ' . $apiKey,
-        ];
+        $zenodoRecId = $this->depositRecord($jsonString, $object, $recordsApiUrl, $apiKey, $isUpdate, $isPublished, $existingZenodoRecId);
+        if (is_array($zenodoRecId)) {
+            // Deposit or update has failed @todo more here?
+            return $zenodoRecId;
+        }
 
-        try {
-            $response = $httpClient->request(
-                'POST',
-                $url,
-                [
-                    'headers' => $headers,
-                    'json' => json_decode($jsonString),
-                ]
-            );
-        } catch (GuzzleException | Exception $e) {
-            return [['plugins.importexport.zenodo.register.error.mdsError', $e->getMessage()]];
+        if ($isUpdate && $isPublished) {
+            // try to re-publish the updated record
+            $republishUpdatedRecord = $this->publishZenodoRecord($object, $zenodoRecId, $recordsApiUrl, $apiKey);
+            if (is_array($republishUpdatedRecord)) {
+                return $republishUpdatedRecord;
+            }
         }
 
-        $responseBody = json_decode($response->getBody());
-        $zenodoRecId = $responseBody->id;
+        // note: can't update files on published records
+        if (!$isUpdate) { // @todo disable file deposit for updates as it will fail if the filename exists already - maybe we should delete all files or the record first in this case
+            $filesDeposit = $this->depositFiles($object, $recordsApiUrl, $apiKey, $zenodoRecId);
+            if (is_array($filesDeposit)) {
+                // File deposit has failed, try to delete the record in Zenodo then return the error
+                $this->deleteRecord($object, $zenodoRecId, $recordsApiUrl, $apiKey);
+                return $filesDeposit;
+            }
+        }
 
-        if (($status = $response->getStatusCode()) != ZENODO_API_DEPOSIT_CREATED) {
-            return [['plugins.importexport.zenodo.register.error.mdsError', $status . ' - ' . $response->getBody()]];
+        // Submit the record to a community
+        $communityId = $this->getCommunityId($context);
+        if ($communityId) {
```

Comment:
Do we need to create review if it is an update of already published record, if we only update the metadata?